### PR TITLE
Improve validation and refactor client

### PR DESCRIPTION
**Pull request by alan-ma**
To be merged into `master` along with changes in `switch_to_faraday`, leaving the `switch_to_faraday` branch since it is being referenced by production systems.

See corresponding BladeRunner PR: https://github.com/Shopify/BladeRunner/pull/2645

# Improve validation and refactor client

### What are you trying to accomplish?

Improve validation to add checks on all attributes, while updating documentation to reflect what is happening. This classifies the attributes into different categories:
- `INPUT_STRING_ATTRIBUTES` - Attributes must be strings
- `INPUT_NUMERIC_ATTRIBUTES` - Attributes must be numeric
- `REQUIRED_ATTRIBUTES` - Attributes must not be empty
- `CONVERT_STRING_ATTRIBUTES` - Attributes will be automatically converted to strings

Make `Transaction` and `Response` immutable after initialization.

Remove the `parse` step on a `Response` - this should be done when the `Response` is initialized.

Add check for `cvv_result` (should be a single letter) and downcase email before hashing (as specified by the Minfraud API).

Allow access to the response attribute Hash via `Response.body` (this returns a frozen Hash).

Allow optional block to be passed during `MinfraudClient.send_transaction` to modify the response Hash during initialization.

Sample usage (updated in the README):

```ruby
# Create a client to handle requests for a specific host
minfraud_client = Minfraud::MinfraudClient.new(
  # Optional fields to customize connection
  host_choice: 'us_east',
  open_timeout: 1,
  idle_timeout: 5,
  read_timeout: 5,
  write_timeout: 5,
  pool_size: 10
)

# Create a transaction with attributes to send
# The transaction will be immutable after creation
transaction = Minfraud::Transaction.new do |t|
  # Required fields
  t.ip = '1.2.3.4'
  t.txn_id = 'Order-1'
  t.license_key = 'minfraud-license-key'
  # Other fields are optional
  t.city = 'richmond'
  t.state = 'virginia'
  t.postal = '12345'
  t.country = 'US' # http://en.wikipedia.org/wiki/ISO_3166-1
  # ...
end

# Send the transaction to minFraud and retrieve the parsed response
# The response will be immutable
response = minfraud_client.send_transaction(
  transaction: transaction,
  # Optional url to customize the request endpoint
  # Default value is `/app/ccv2r`
  url: '/app/ccv2r'
)

# All attributes from minFraud's response will be available on the response object
# as a snake-cased attribute. For example, if you want the `proxyScore` field
# from minFraud, you can get it with `#proxy_score`.

response.risk_score
# => 3.48

response.queries_remaining
# => 1000

# A warning may be present in the `#err` attribute
response.err
# => nil

# The attributes can also be retrieved as a Hash
# Note that this is a frozen Hash
response_hash = response.body # frozen Hash
response_hash = response.body.dup # mutable duplicate of Hash

response[:risk_score]
response[:queries_remaining]
response[:err]
```

### Why did you choose this approach?

Followed type requirements in the [Minfraud Legacy API](https://dev.maxmind.com/minfraud/minfraud-legacy/), as well as what we're [currently passing in from core](https://github.com/Shopify/shopify/blob/d41b279f2c4a3c1b255c7d79bfeba0f38b451323/components/reviews/app/models/fraud_analysis/minfraud_analysis_attributes.rb#L11-L40).

Added `CONVERT_STRING_ATTRIBUTES` with `:shop_id` since we currently pass this as an integer, but Minfraud actually expects a string. This makes the typing of `:shop_id` more explicit.

Froze `Transaction` and `Response` after initialization to use them as immutable value objects.

### What should reviewers focus on?

Should converting the attributes in `CONVERT_STRING_ATTRIBUTES` happen in `validate_attributes` or should we do this more explicitly in another method?

### How are you going to verify this change?

Updated unit tests and tophatted as part of https://github.com/Shopify/BladeRunner/pull/2645.

# Switch to Faraday

### What are you trying to accomplish?

Switch to using [Faraday](https://github.com/lostisland/faraday) to make requests to MinFraud. This makes use of [Net::HTTP::Persistent](https://github.com/drbrain/net-http-persistent) to share connections in a connection pool. The goal is to increase the robustness of the client and as a bonus, speed up response times.

Added `Minfraud::MinfraudClient` as a way to send requests. The `send_transaction` method takes in a `Minfraud::Transaction` and returns a `Minfraud::Response`.

Removed configuration requirements for the `Minfraud` module.

Made `license_key` a required field on `Minfraud::Transaction`.

Updated version to `1.0.0`.

Sample usage (updated in the README):

```ruby
minfraud_client = Minfraud::MinfraudClient.new(
  # Optional fields to customize connection
  host_choice: 'us_east',
  open_timeout: 1,
  idle_timeout: 5,
  read_timeout: 5,
  write_timeout: 5,
  pool_size: 10
)

transaction = Minfraud::Transaction.new do |t|
  # Required fields
  # Other fields listed later in documentation are optional
  t.ip = '1.2.3.4'
  t.city = 'richmond'
  t.state = 'virginia'
  t.postal = '12345'
  t.country = 'US' # http://en.wikipedia.org/wiki/ISO_3166-1
  t.txn_id = 'Order-1'
  t.license_key = 'minfraud-license-key'
  # ...
end

response = minfraud_client.send_transaction(
  transaction: transaction,
  # Optional field to customize connection
  url: '/app/ccv2r'
)

response.parse # parses body to create hash

response.dig(:risk_score)
# => 3.48
```

### Why did you choose this approach?

This decouples the request logic from the `Minfraud::Transaction` object. This allows `Minfraud::Transaction`s to purely be containers for the data being passed around, and makes the flow simpler.

Removing the configuration step means that all connection customization will now come from the `Minfraud::MinfraudClient` and `Minfraud::Transaction` objects, making it more clear what's going on.

### What should reviewers focus on?

SSL verification was previously off. [As with the discussion here](https://github.com/Shopify/shopify/pull/255784/files#r466110438), I'm not sure if we should be bypassing verification. I opted to leave verification settings to Faraday defaults, which does have SSL verification.

### How are you going to verify this change?

Updated unit tests.

Tested locally and in staging by making requests in BladeRunner. More details on testing can be found in the [BladeRunner-side PR](https://github.com/Shopify/BladeRunner/pull/2623).

### Before you deploy

 - [x] [:tophat:](https://vault.shopify.com/developers/Tophatting)'d
 - [x] Safe to rollback?

@Shopify/riskeng

### CODE REVIEW

**Review by dharkael — COMMENTED** [review]
This is very complete and very well done. Great work!

**Review by dharkael — APPROVED** [review]

File: lib/minfraud/minfraud_client.rb
Comment (Arafatk) [inline lib/minfraud/minfraud_client.rb:48]: [Faraday code ](https://github.com/lostisland/faraday/blob/e331955fd87648f4c6a6488cda180a68a734e8a7/lib/faraday/options/env.rb#L18-L20)

File: lib/minfraud/transaction.rb
Comment (Arafatk) [inline lib/minfraud/transaction.rb:121]: Here the Numeric type is used as an all encompassing type of all attributes that require the validation of a number.      

The actual type of `amount` attribute is `BigDecimal` and `BigDecimal` inherits from Numeric class.

Comment (alan-ma) [reply lib/minfraud/transaction.rb:121]: Yep! Should this validation ensure that a `BigDecimal` is needed? Other users of this client may pass in something else, so I went with `Numeric`. (I also included a test to make sure `BigDecimal` is properly accepted)

**Comment by Arafatk** [comment]
NOTE: 
1. This PR will not be merged until codechill. The group consensus is that this PR can be merged into core only after the the [BFCM Code Chill](https://vault.shopify.io/pages/7023-BFCM-2020). Since alan won't be here, I will take the ownership of the deploy.             

2. This PR will be merged directly into master and then Bladerunner will be shifted back to master.         

Once this PR is merged https://github.com/Shopify/shopify/pull/270368 can switch to using master from MinfraudRuby.

File: README.md
Comment (alan-ma) [inline README.md:80]: This was added functionality to allow customization that BladeRunner requires

**Review by Arafatk — APPROVED** [review]

**Review by MattGingerich — APPROVED** [review]
Great work on this. 👍 🎉

File: lib/minfraud/minfraud_client.rb
Comment (erickadbay) [inline lib/minfraud/minfraud_client.rb:77]: Does a `Response` care about the block? Why not call the block from within this method?

File: lib/minfraud/minfraud_client.rb
Comment (erickadbay) [inline lib/minfraud/minfraud_client.rb:70]: Is this modification on the response itself? Isn't it immutable?

File: lib/minfraud/transaction.rb
Comment (erickadbay) [inline lib/minfraud/transaction.rb:89]: [`Object#blank?`](https://apidock.com/rails/Object/blank%3F) might be able to help here 🙂

File: lib/minfraud/transaction.rb
Comment (erickadbay) [inline lib/minfraud/transaction.rb:99]: Isn't `cvv_result` already being validated by validating `INPUT_STRING_ATTRIBUTES`?

File: lib/minfraud/transaction.rb
Comment (erickadbay) [inline lib/minfraud/transaction.rb:142]: `@return [void]` might be more representative of what the method does since all it does is assigning an attribute 🙂

**Review by erickadbay — COMMENTED** [review]
Some minor comments but looks great!

File: lib/minfraud/transaction.rb
Comment (alan-ma) [reply lib/minfraud/transaction.rb:89]: apparently #blank? and #present? are Rails methods, so they can't be used here 😅

File: lib/minfraud/transaction.rb
Comment (alan-ma) [reply lib/minfraud/transaction.rb:99]: That's an interesting assumption - `INPUT_STRING_ATTRIBUTES` only validates that the attributes are strings, without any additional checks.

The `cvv_result` is one attribute that is being validated more strictly (it should be a single character), and more checks may be added in the future.

Should I add an array of `SINGLE_CHARACTER_ATTRIBUTES` and verify in a similar manner instead of doing this with `validate_cvv`?

File: lib/minfraud/minfraud_client.rb
Comment (alan-ma) [reply lib/minfraud/minfraud_client.rb:70]: This happens during the initialization of the Response (before it becomes immutable) - updated the comment to make this more clear!

File: lib/minfraud/minfraud_client.rb
Comment (alan-ma) [reply lib/minfraud/minfraud_client.rb:77]: The block is passed in during initialization so it can be called before it freezes itself - this way, all Responses are guaranteed to be frozen, but it also allows for modification of the attributes during initialization

Added a comment to make this more clear

File: lib/minfraud/transaction.rb
Comment (erickadbay) [reply lib/minfraud/transaction.rb:89]: Ahh of course 🤦

**Review by dharkael — APPROVED** [review]

**Comment by Arafatk** [comment]
@erickadbay On my end, It seems like all your questions have a good response.      
 
Feel free to mention any other questions/suggestions from this PR so that I can move forward with the merging of these set of PRs.

**Review by erickadbay — APPROVED** [review]

**Pull request merged**
